### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by willeslau**
Initial implementation of generating proofs. Implemented the trait, will implement the Forest adaptor next

### CODE REVIEW

File: filecoindot_proofs/src/traits.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    fn path_to_key<S: BlockStore<K, V, H, Self>>(
+        &self,
+        hash_bits: &mut H,
+        k: &K,
+        path: &mut Vec<Vec<u8>>,
+        bit_width: u8,
+        store: &S,
+    ) -> Result<bool, Error>
+    where
+        Self: Sized;
```
[the remark this answers is not in the document]
Comment (willeslau) [reply filecoindot_proofs/src/traits.rs:33]: yep and the path will contain the full path from the root to the node

**Review by mattsse — APPROVED** [review]
awesome!

File: filecoindot_proofs/src/verify.rs [hunk 2]
```diff
@@ -0,0 +1,46 @@
+use cid::Cid;
+use crate::errors::Error;
+
+pub fn verify(proof: Vec<Vec<u8>>, cid: &Cid) -> Result<(), Error> {
+    let bytes = cid.to_bytes();
+    match proof.iter().find(|p| **p == bytes) {
+        Some(_) => Ok(()),
+        None => Err(Error::VerificationFailed)
+    }
+}
```
Comment (mattsse) [inline filecoindot_proofs/src/verify.rs:10]: this only verifies that the cid is included the proof somewhere,
we need to also check that the root matches, so a verify function would look like `fn verify (root, proof, cid)`

**Review by mattsse — CHANGES_REQUESTED** [review]

File: filecoindot_proofs/src/verify.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+pub fn verify(root: &Cid, proof: Vec<Vec<u8>>, cid: &Cid) -> Result<(), Error> {
+    if proof.is_empty() || proof[0] != root.to_bytes() {
+        return Err(Error::VerificationFailed);
+    }
+
+    let bytes = cid.to_bytes();
+    match proof.iter().find(|p| **p == bytes) {
+        Some(_) => Ok(()),
+        None => Err(Error::VerificationFailed),
+    }
```
Comment (mattsse) [inline filecoindot_proofs/src/verify.rs:17]: this only if the proof node is included in the list,
we need to ensure that the proof is actually correct, by traversing the whole proof nodes again and check that all hashes match as expected

**Review by mattsse — CHANGES_REQUESTED** [review]

File: filecoindot_proofs/src/verify.rs (hunk 3 above)
Comment (willeslau) [reply filecoindot_proofs/src/verify.rs:17]: Ok ok, I'll add a struct that handles this

File: filecoindot_proofs/src/forest_amt_adaptor.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+use serde::{Deserializer, Serialize, Serializer};
+use serde::de::Error as SerdeError;
+use crate::amt::nodes_for_height;
+
+const DEFAULT_BIT_WIDTH: usize = 3;
+
+impl From<ipld_amt::Error> for Error {
+    fn from(error: ipld_amt::Error) -> Self {
+        let error_str = format!("forest_db error: {:?}", error);
+        Error::Other(error_str)
```
Comment (ivanceras) [inline filecoindot_proofs/src/forest_amt_adaptor.rs:18]: This could just `Error::ForrestDBErr(error)`

**Review by ivanceras — CHANGES_REQUESTED** [review]
The error for forrest/ipld specific error needs to be a variant of the enum `Error`, future code might have to deal with specific error.

**Review by ivanceras — APPROVED** [review]
lgtm!

**Review by mattsse — CHANGES_REQUESTED** [review]
some clippy fixes and lints, otherwise lgtm!

**Review by mattsse — APPROVED** [review]
great stuff!

**Pull request merged**